### DISCUSSION TITLE
4:4:4 color mode

### DISCUSSION THREAD

**Discussion by test2a** [rustdesk/rustdesk#4952] [Q&A] [answered]
parsec seems to support 4.4.4 color mode so the same probably would be beneficial for rustdesk users as well

**Comment by vsuvarna** [comment]
This is a key selling point for the Media and Entertainment industry or content creation in general.

**Reply by rustdesk** [reply]
@21pages

**Reply by ruchit-bhatt** [reply]
+1000
Essential feature for M&E

**Comment by kevin-wijnen** [comment]
> parsec seems to support 4.4.4 color mode so the same probably would be beneficial for rustdesk users as well

Even for consumer use, this would be an amazing feature. My current use case is to stream a headless Mac to another PC. And the limitations of Parsec's colour range (free plan has the same limit as rustdesk, except for paid plans) are definitely making coding remote (with iOS Simulator) difficult due to red text etc.

**Comment by 21pages** [comment]
@test2a @vsuvarna @ruchit-bhatt @kevin-wijnen 

I make a pull request https://github.com/rustdesk/rustdesk/pull/6229, but It's hard for me to see the quality difference between 420 and 444, but I can ensure the 444 is converted from bgra directly and converted to bgra directly, could you try it and test if 444 makes some differences?
https://github.com/21pages/rustdesk/releases/download/nightly/rustdesk-1.2.4-x86_64-444.exe, other platforms are also available, https://github.com/21pages/rustdesk/releases/tag/nightly

**Reply by vsuvarna** [reply]
Thank you so much for this @21pages. This is a huge update, especially for people working in the creative industry. I've installed the nightly version as advised and I can see the 4:4:4 setting. I'll test this out over the coming weeks and report back if there are any issues. Thank you once again for your hard work.

**Reply by kevin-wijnen** [reply]
Thank you very much for the feature @21pages ! I'll test it in the coming days.

Personally, I'd recommend testing the chroma subsampling with this testing method of RTINGS: https://www.rtings.com/tv/learn/chroma-subsampling

It has been used in the past by Digital Foundry to compare streaming applications regarding 4:2:0 and 4:4:4 modes:
https://youtu.be/BKd6vTE3xgk?t=966

Also, thank you for providing the builds for multiple platforms as well!

**Reply by 21pages** [reply]
You can try vp9 first, it seems there is a difference in vp9 but not in av1.

**Reply by kevin-wijnen** [reply]
> You can try vp9 first, it seems there is a difference in vp9 but not in av1.

@21pages It may be related to the hardware? AV1 encoding (and decoding) requires quite modern hardware to accomplish. On the NVIDIA GPU side, starting with RTX 3xxx series their cards can decode AV1, while only RTX 4xxx series can also encode AV1 video.

Don't know the AMD/Intel GPU limitations out of the top of my head.

**Reply by 21pages** [reply]
Currently, AV1 is purely software encoding. If 444 is effective, h264/h265 support may be added.
New version AMD support AV1.

**Comment by 21pages** [comment, marked as answer]
444 works at the pixel level at least, this is my test.

1. Create an image that is the same size as the screen (2880x1800 in my pc). When the length and width are divisible by 5, the pixel is  blue, otherwise it is black.
2. Adjust the screen zoom to 100%, maximize the above picture, and use rustdesk's vp9_420, vp9_444, av1_420, av1_444 to record the video.
3. Use vlc to play the video in full screen and take a screenshot using vlc when the picture above shows in full screen state
4. Use photoshop to zoom in to 3200%, and you can see that 444 is a single pixel of pure blue, and 420 is 4 non-standard blues.

Here are pictures of all the steps: https://github.com/21pages/test/releases/download/test/444.zip

Here are the pictures in photoshop:

![ps_vp9_420](https://github.com/rustdesk/rustdesk/assets/14891774/3c759c73-a0b5-493d-8606-35849b4ede2b)

![ps_vp9_444](https://github.com/rustdesk/rustdesk/assets/14891774/1c6da493-9c9f-40e4-9464-4d698181f61b)

![ps_av1_420](https://github.com/rustdesk/rustdesk/assets/14891774/ff6b34f6-d699-46de-a017-7ee0c175ab3a)

![ps_av1_444](https://github.com/rustdesk/rustdesk/assets/14891774/cfef1e0b-ecd9-470f-be74-0ec2b00f6435)

You can also see the difference of the full size screenshot pictures without zoom, 420 is almost black, 444 is more blue.

**Comment by ruchit-bhatt** [comment]
@21pages support for 4:4:4 also added to H:264 and H:265 NVENC on nvidia GPU ?

**Reply by 21pages** [reply]
not yet, may be later

**Comment by deep-soft** [comment]
Tested Chroma 4:4:4 on:
windows  7 host - windows 10 client, VP9 and AV1
windows  7 host - windows 11 client, VP9 and AV1
windows 10 host - windows 11 client, VP9 and AV1

I think it will be useful if Quality Monitor will display, next to Codec, also if hardware or software compression it's used.
Something like this:
AV1 hw / AV1 sw
VP9 hw / VP9 sw

**Comment by vsuvarna** [comment]
Tested 4:4:4 mode today with the following settings:

- default codec: VP9
- other default options: Truecolor (4:4:4)
- default image quality: good image quality

All hosts had these exact settings. We tried a Win7 host and Win10 client also vice versa. We used the test pattern suggested by @kevin-wijnen. We also had an instance of google meet running along side for communication and comparison. There was definitely a huge difference in colour perforamance between google meet and rustdesk. We verified the results by comparing to the downloaded png. The difference is stark and easily noticeable and rustdesk PQ was fantastic.

Few hiccups we had along the way:
- Rustdesk prevented me from taking screenshots. I could have used it to verify our findings in photoshop.
- Another client using Mac OS v10.10.5 was not able to install the supplied rustdesk nightly due to it being an older version. If it is possible, seeing a wider coverage of older devices could be useful to many.
- Some new users found the setting process a little complicated. I know it is just 3 options, but if there was a 4:4:4 toggle switch that auto selects the relevant features, it would be useful for non tech savvy people who don't know what any of the options mean.

Suggestions for the future:
This is a great first step that will definitely bring rustdesk to the limelight of creative professionals. However if we want to reach feature parity with the industry leader of colour accurate streams - Colorfront, then we should consider introducing the following supplementing features for 4:4:4 mode:

- 4:4:4 mode ✓
- 10bit colour
- HEVC / AV1 codec support
- SRT protocol
- AES-256 encryption (optional for now)
- Watermarking capability

If we have these features, rustdesk can be promoted as a drop in replacement to colorfront and it will be useful for ad agencies, post houses, vfx studios, editing studios and so on. I know its still early days for this feature, but it has immense potential to be a game changer for the creative industry. Thank you once again for your hard work and for implementing this feature request.

**Reply by alatteri** [reply]
What you describe with ColorFront Streaming Server is totally different that RustDesk, and really doesn't apply here at all.

**Comment by vsuvarna** [comment]


**Reply by 21pages** [reply]
https://github.com/rustdesk/rustdesk/releases/tag/nightly
